shuttlepro.1
.\" Automatically generated by Pandoc 2.2.3.2
.\"
.TH "shuttlepro" "1" "" "" ""
.hy
.SH Name
.PP
shuttlepro \[en] translate input from the Contour Design Shuttle devices
.SH Synopsis
.PP
shuttlepro [\-h] [\-o] [\-p] [\-j \f[I]name\f[]] [\-r \f[I]rcfile\f[]]
[\-d[rskj]] [\f[I]device\f[]]
.SH Options
.TP
.B \-h
Print a short help message.
.RS
.RE
.TP
.B \-o
Enable MIDI output.
.RS
.RE
.TP
.B \-p
Enable hotplugging support.
.RS
.RE
.TP
.B \-j \f[I]name\f[]
Set the Jack client name.
Default: \[lq]shuttlepro\[rq].
.RS
.RE
.TP
.B \-r \f[I]rcfile\f[]
Set the configuration file name.
Default: taken from the SHUTTLE_CONFIG_FILE environment variable if it
exists, or ~/.shuttlerc if it exists, /etc/shuttlerc otherwise.
.RS
.RE
.TP
.B \-d[rskj]
Enable various debugging options: r = regex (print matched translation
sections), s = strokes (print the parsed configuration file in a
human\-readable format), k = keys (print executed translations), j =
jack (additional Jack debugging output).
Just \f[C]\-d\f[] enables all debugging options.
.RS
.RE
.SH Description
.PP
The Contour Design Shuttle devices are specialized input devices mainly
used with audio and video editing software.
They offer three different kinds of controls: a jog wheel (an endless
rotary encoder), a shuttle wheel surrounding the jog wheel which
automatically snaps back into its center position, and a collection of
buttons arranged around the jog and shuttle wheels.
.PP
There are various versions of these devices.
The current line of products from Contour Design consists of the
ShuttlePRO v2 and the ShuttleXpress.
These and a few older versions should be recognized by the shuttlepro
program without any further ado.
However, there are also some rebranded versions which might not be
detected automatically, in which case you'll have to specify the device
name on the command line.
.PP
The shuttlepro program enables you to use these devices on Linux.
It will work with any application taking X11 keyboard and mouse or MIDI
input, without requiring any special support from the application.
The program translates input events (button presses, jog and shuttle
wheel movements) into X keystrokes, mouse button presses, scroll wheel
events, or, as an option, MIDI output.
It does this by matching the \f[C]WM_CLASS\f[] and \f[C]WM_NAME\f[]
properties of the window that has the keyboard focus against the regular
expressions for each application section in its configuration
(shuttlerc) file.
If a regex matches, the corresponding set of translations is used.
If a matching section cannot be found, or if it doesn't define a
suitable translation, the program falls back to a set of default
translations at the end of the file, if available.
.PP
The shuttlerc file is just an ordinary text file which you can edit to
configure the program for your applications.
A sample configuration file is included, see example.shuttlerc in the
sources.
This also gets installed as a system\-wide configuration file, so that
the program works out of the box with any of the preconfigured
applications.
At the time of this writing, there are ready\-to\-use translations for
some popular video editors and a basic Mackie emulation which should
work with most DAW (digital audio workstation) programs.
Your contributions are appreciated, so if you have any interesting
application configurations to share, please submit them at the project
page (https://github.com/agraef/ShuttlePRO)!
.SS Installation
.PP
First, make sure that you have the required dependencies installed.
The program needs a few X11 libraries and Jack (http://jackaudio.org/);
the latter is only required if you plan to utilize the MIDI support.
And of course you need GNU make and gcc (the GNU C compiler).
On Ubuntu and other Debian\-based systems you should be able to get
everything that's needed by running this command:
.IP
.nf
\f[C]
sudo\ apt\ install\ build\-essential\ libx11\-dev\ libxtst\-dev\ libjack\-dev
\f[]
.fi
.PP
Then just run \f[C]make\f[] and \f[C]sudo\ make\ install\f[].
This installs the example.shuttlerc file as /etc/shuttlerc, and the
shuttlepro program and the manual page in the default install location.
Usually this will be under /usr/local, but the installation prefix can
be changed with the \f[C]prefix\f[] variable in the Makefile.
Also, package maintainers can use the \f[C]DESTDIR\f[] variable as usual
to install into a staging directory for packaging purposes.
.PP
The program will automatically be built with Jack MIDI support if the
Jack development files are detected at compile time.
(If you do have Jack installed, but you still want to build a Jack\-less
version of the program for some reason, you can do that by running
\f[C]make\ JACK=\f[] instead of just \f[C]make\f[].)
.PP
Finally, if you also want hotplugging support, you can run
\f[C]sudo\ make\ install\-udev\f[] to have the corresponding udev rules
and helper script installed.
(If you specified the \f[C]prefix\f[] variable during
\f[C]make\ install\f[], the same value should be used here, so that the
proper absolute path to the shuttlepro program gets inserted into the
helper script.) However, we recommend that you first go without this,
until you have everything set up as needed, and even then you should
first read the \f[I]Hotplugging\f[] section at the end of this manual
and note the caveats concerning this feature.
.SH Configuration File
.PP
After installation the system\-wide default configuration file will be
in /etc/shuttlerc, where the program will be able to find it.
We recommend copying this file to your home directory, renaming it to
\&.shuttlerc:
.IP
.nf
\f[C]
cp\ /etc/shuttlerc\ ~/.shuttlerc
\f[]
.fi
.PP
The ~/.shuttlerc file, if it exists, takes priority over /etc/shuttlerc,
so it becomes your personal shuttlepro configuration.
You can edit this file as you see fit, in order to customize the
configuration for the applications that you use.
.PP
The program automatically reloads the configuration file whenever it
notices that the file has been changed.
Thus you can edit the file while the program keeps running, and have the
changes take effect immediately without having to restart the program.
The program also has a bunch of debugging options which let you to see
exactly how your translations are being processed, which is very helpful
when testing out new translations.
.SH Usage
.PP
The shuttlepro program is a command line application, so usually you run
it from the terminal.
However, once you've set up everything to your liking, you can also
launch it from your Jack session manager (see \f[I]MIDI Output\f[]
below), or from your desktop environment's startup files.
Hotplugging support using the Linux udev system is also available;
please check the \f[I]Hotplugging\f[] section below for details.
.PP
Before you can use the program, you have to make sure that you can
access the device.
On modern Linux systems, becoming a member of the \f[C]input\f[] group
should be all that is needed:
.IP
.nf
\f[C]
sudo\ usermod\ \-a\ \-G\ input\ username
\f[]
.fi
.PP
Log out and in again, and you should be set.
(Another possibility is to use the hotplugging feature which will set
the permissions of the device so that an ordinary user has access even
without being member of the \f[C]input\f[] group; please check the
\f[I]Hotplugging\f[] section for details.)
.PP
Now make sure that your Shuttle device is connected, and try running
\f[C]shuttlepro\f[] from the command line (without any arguments).
The program should hopefully detect your device and print something
like:
.IP
.nf
\f[C]
shuttlepro:\ found\ shuttle\ device:
/dev/input/by\-id/usb\-Contour_Design_ShuttleXpress\-event\-if00
\f[]
.fi
.PP
(The precise name of the device will differ, depending on the version of
the device that you have.
E.g., the output above indicates that a Shuttle Xpress was found.)
.PP
If the program fails to find your device, you'll have to locate it
yourself and specify the absolute pathname to it on the command line.
On modern Linux systems, there should be an entry under
/dev/input/by\-id for it, which is simply a symbolic link to some device
node under /dev/input.
Look for devices having \[lq]Contour_Design\[rq], \[lq]CAVS\[rq] or
\[lq]Shuttle\[rq] in their name.
.PP
Naming the device on the command line will also be necessary if you have
multiple Shuttle devices.
In this case you may want to run a separate instance of shuttlepro for
each of them (possibly with different configurations, using the
\f[C]\-r\f[] option).
.PP
If your device was found, you should be able to operate it now and have,
e.g., the terminal window in which you launched the program scroll and
execute mouse clicks if you move the jog wheel and press the three
center buttons on the device.
When you're finished, terminate the program by typing Ctrl+C in the
terminal window where you launched it.
.PP
Note that once up, the program will just keep running, even if the
device gets unplugged, in which case an error message will be printed
and the program will keep trying to reopen the device until you
interrupt it.
However, there is an alternative mode available with the \f[C]\-p\f[]
option, which makes sure that the program exits as soon as the device
becomes unavailable.
This is also used, in particular, when using the udev hotplugging
facility mentioned above.
.PP
The default \[lq]mouse emulation mode\[rq] is actually configured in the
\f[C][Default]\f[] section near the end of the distributed shuttlerc
file, which reads as follows:
.IP
.nf
\f[C]
[Default]
\ K6\ XK_Button_1
\ K7\ XK_Button_2
\ K8\ XK_Button_3
\ JL\ XK_Scroll_Up
\ JR\ XK_Scroll_Down
\f[]
.fi
.PP
As you can see, the buttons denoted \f[C]K6\f[], \f[C]K7\f[] and
\f[C]K8\f[] (which are the three buttons right above the jog wheel, see
the comments at the beginning of the shuttlerc file for a description of
the button layout) are mapped to the corresponding mouse buttons, and
rotating the jog wheel to the left (\f[C]JL\f[]) and right (\f[C]JR\f[])
emulates the scroll wheel, scrolling up and down, respectively.
(Besides these mouse actions, you can also bind input events to
arbitrary sequences of key strokes, so operating the functions of any
application that is well\-equipped with keyboard shortcuts should in
most cases be a piece of cake.
Have a look at the other configuration entries to see how this is done.)
.PP
One useful feature is that you can invoke the program with various
debugging options to get more verbose output as the program recognizes
events from the device and translates them to corresponding mouse
actions or key presses.
E.g., try running \f[C]shuttlepro\ \-drk\f[] to have the program print
the recognized configuration sections and translations as they are
executed.
For instance, here is what the program may print in the terminal if you
move the jog wheel one tick to the right (\f[C]JR\f[]), then left
(\f[C]JL\f[]), and finally press the leftmost of the three buttons
(\f[C]K6\f[]):
.IP
.nf
\f[C]
$\ shuttlepro\ \-drk
shuttlepro:\ found\ shuttle\ device:
/dev/input/by\-id/usb\-Contour_Design_ShuttleXpress\-event\-if00
Loading\ configuration:\ /home/foo/.shuttlerc
translation:\ Default\ for\ ShuttlePRO\ :\ bash\ (class\ konsole)
JR:\ XK_Scroll_Down/D\ XK_Scroll_Down/U\ 
JL:\ XK_Scroll_Up/D\ XK_Scroll_Up/U\ 
K5[D]:\ XK_Button_1/D\ 
K5[U]:\ XK_Button_1/U\ 
\f[]
.fi
.PP
We strongly recommend using the debugging options when editing the
translations, so that you can see exactly what's going on if your
translations don't appear to work correctly.
The \f[C]\-d\f[] option can be combined with various option characters
to choose exactly which kinds of debugging output you want; \f[C]r\f[]
(\[lq]regex\[rq]) prints the matched translation section (if any) along
with the window name and class of the focused window; \f[C]s\f[]
(\[lq]strokes\[rq]) prints the parsed contents of the configuration file
in a human\-readable form whenever the file is loaded; \f[C]k\f[]
(\[lq]keys\[rq]) shows the recognized translations as the program
executes them, in the same format as \f[C]s\f[]; and \f[C]j\f[] adds
some debugging output from the Jack driver.
You can also just use \f[C]\-d\f[] to enable all debugging output.
Most of these options are also available as directives in the shuttlerc
file; please check the distributed example.shuttlerc for details.
.SS MIDI Output
.PP
If the shuttlepro program was built with Jack MIDI support, it can also
be used to translate input from the Shuttle device to corresponding MIDI
messages rather than key presses.
This is useful if you want to hook up the device to any kind of
MIDI\-capable program, such as software synthesizers or digital audio
workstation (DAW) programs like Ardour (https://ardour.org/).
.PP
You need to run the program as \f[C]shuttlepro\ \-o\f[] to enable MIDI
output at run time.
This will start up Jack (if it is not already running) and create a Jack
client named \f[C]shuttlepro\f[] with a single MIDI output port which
can then be connected to the MIDI inputs of other programs.
The Jack client name can also be changed with the \f[C]\-j\f[] option,
which is useful if you're running multiple instances of the program with
different Shuttle devices (possibly using different configurations).
.PP
We recommend using a Jack front\-end and patchbay program like
QjackCtl (https://qjackctl.sourceforge.io/) to manage Jack and to set up
the MIDI connections.
In QjackCtl's setup, make sure that you have selected \f[C]seq\f[] as
the MIDI driver.
This exposes the ALSA sequencer ports of non\-Jack ALSA MIDI
applications as Jack MIDI ports, so that they can easily be connected to
shuttlepro.
(We're assuming that you're using Jack1 here.
Jack2 works in a very similar way, but may require some more fiddling;
in particular, you may have to use
a2jmidid (http://repo.or.cz/a2jmidid.git) as a separate ALSA\-Jack MIDI
bridge in order to have the ALSA MIDI devices show properly as Jack MIDI
devices.)
.PP
The shuttlepro program also supports Jack session management, which
makes it possible to record the options the program was invoked with
along with the MIDI connections.
This feature can be used with any Jack session management software.
Specifically, QjackCtl has its own built\-in Jack session manager which
is available in its Session dialog.
To use this, launch shuttlepro and any other Jack applications you want
to have in the session, use QjackCtl to set up all the connections as
needed, and then hit the \[lq]Save\[rq] button in the Session dialog to
have the session recorded.
Now, at any later time you can relaunch the same session with the

shuttlepro.c

/*

 Contour ShuttlePro v2 interface

 Copyright 2013 Eric Messick (FixedImagePhoto.com/Contact)

 Copyright 2018 Albert Graef <aggraef@gmail.com>, various improvements

 Based on a version (c) 2006 Trammell Hudson <hudson@osresearch.net>

 which was in turn

 Based heavily on code by Arendt David <admin@prnet.org>

*/

#include "shuttle.h"
#if HAVE_JACK
#include "jackdriver.h"
#endif

typedef struct input_event EV;

extern int debug_regex;
extern translation *default_translation;

unsigned short jogvalue = 0xffff;
int shuttlevalue = 0xffff;
struct timeval last_shuttle;
int need_synthetic_shuttle;
Display *display;

#if HAVE_JACK
JACK_SEQ seq;
#endif
int enable_jack = 0, debug_jack = 0;

void
initdisplay(void)
{
  int event, error, major, minor;

  display = XOpenDisplay(0);
  if (!display) {
    fprintf(stderr, "unable to open X display\n");
    exit(1);
  }
  if (!XTestQueryExtension(display, &event, &error, &major, &minor)) {
    fprintf(stderr, "Xtest extensions not supported\n");
    XCloseDisplay(display);
    exit(1);
  }
}

void
send_button(unsigned int button, int press)
{
  XTestFakeButtonEvent(display, button, press ? True : False, DELAY);
}

void
send_key(KeySym key, int press)
{
  KeyCode keycode;

  if (key >= XK_Button_1 && key <= XK_Scroll_Down) {
    send_button((unsigned int)key - XK_Button_0, press);
    return;
  }
  keycode = XKeysymToKeycode(display, key);
  XTestFakeKeyEvent(display, keycode, press ? True : False, DELAY);
}

#if HAVE_JACK
// cached controller and pitch bend values
static int ccvalue[16][128];
static int pbvalue[16] =
  {8192, 8192, 8192, 8192, 8192, 8192, 8192, 8192,
   8192, 8192, 8192, 8192, 8192, 8192, 8192, 8192};

void
send_midi(int status, int data, int incr, int kjs, int index)
{
  if (!enable_jack) return; // MIDI support not enabled
  uint8_t msg[3];
  int chan = status & 0x0f;
  msg[0] = status;
  msg[1] = data;
  switch (status & 0xf0) {
  case 0x90:
    // send a note on (with maximum velocity) for KJS_KEY_DOWN, note off for
    // KJS_KEY_UP, or both for non-key input
    if (kjs == KJS_KEY_DOWN)
      msg[2] = 127;
    else if (kjs == KJS_KEY_UP)
      msg[2] = 0;
    else {
      msg[2] = 127;
      queue_midi(&seq, msg);
      msg[2] = 0;
    }
    break;
  case 0xb0:
    switch (kjs) {
      // for key events, send a controller value of 127 (KJS_KEY_DOWN) or 0
      // (KJS_KEY_UP), similar to how note events are handled
    case KJS_KEY_DOWN:
      msg[2] = 127;
      break;
    case KJS_KEY_UP:
      msg[2] = 0;
      break;
      // for the jog wheel, increment (index==1) or decrement (index==0) the
      // current value, clamping it to the 0..127 data byte range
    case KJS_JOG:
      if (incr) {
	// incremental controller, simply spit out a relative sign bit value
	msg[2] = index?1:65;
      } else if (index) {
	if (ccvalue[chan][data] >= 127) return;
	msg[2] = ++ccvalue[chan][data];
      } else {
	if (ccvalue[chan][data] == 0) return;
	msg[2] = --ccvalue[chan][data];
      }
      break;
      // the shuttle is handled similarly in relative mode, but scaled from
      // 0..14 to the 0..127 data byte range (approximately)
    case KJS_SHUTTLE_INCR:
      if (index) {
	if (ccvalue[chan][data] >= 127) return;
	ccvalue[chan][data] += 9;
	if (ccvalue[chan][data] > 127) ccvalue[chan][data] = 127;
      } else {
	if (ccvalue[chan][data] == 0) return;
	ccvalue[chan][data] -= 9;
	if (ccvalue[chan][data] < 0) ccvalue[chan][data] = 0;
      }
      msg[2] = ccvalue[chan][data];
      break;
      // absolute shuttle (index gives value), scaled in the same fashion
    case KJS_SHUTTLE:
      msg[2] = index * 9;
      break;
    default:
      // this can't happen
      return;
    }
    break;
  case 0xe0: {
    // pitch bends are treated similarly to a controller, but with a 14 bit
    // range (0..16383, with 8192 being the center value)
    int pbval = 0;
    switch (kjs) {
    case KJS_KEY_DOWN:
      pbval = 16383;
      break;
    case KJS_KEY_UP:
      // we use 8192 (center) as the "home" (a.k.a. "off") value, so the pitch
      // will only bend up, never down below the center value
      pbval = 8192;
      break;
    case KJS_JOG:
      if (index) {
	if (pbvalue[chan] >= 16383) return;
	pbvalue[chan] += 128;
	if (pbvalue[chan] > 16383) pbvalue[chan] = 16383;
      } else {
	if (pbvalue[chan] == 0) return;
	pbvalue[chan] -= 128;
	if (pbvalue[chan] < 0) pbvalue[chan] = 0;
      }
      pbval = pbvalue[chan];
      break;
    case KJS_SHUTTLE_INCR:
      if (index) {
	if (pbvalue[chan] >= 16383) return;
	pbvalue[chan] += 1170;
	if (pbvalue[chan] > 16383) pbvalue[chan] = 16383;
      } else {
	if (pbvalue[chan] == 0) return;
	pbvalue[chan] -= 1170;
	if (pbvalue[chan] < 0) pbvalue[chan] = 0;
      }
      pbval = pbvalue[chan];
      break;
    case KJS_SHUTTLE:
      // make sure that we get a proper center value of 8192
      pbval = (index-7)*1170;
      pbval += 8192;
      break;
    default:
      return;
    }
    // the result is a 14 bit value which gets encoded as a combination of two
    // 7 bit values which become the data bytes of the message
    msg[1] = pbval & 0x7f; // LSB (lower 7 bits)
    msg[2] = pbval >> 7;   // MSB (upper 7 bits)
    break;
  }
  case 0xc0:
    // just send the message (once for key presses)
    if (kjs == KJS_KEY_UP) return;
    break;
  default:
    return;
  }
  queue_midi(&seq, msg);
}
#endif

stroke *
fetch_stroke(translation *tr, int kjs, int index)
{
  if (tr != NULL) {
    switch (kjs) {
    case KJS_SHUTTLE:
      return tr->shuttle[index];
    case KJS_SHUTTLE_INCR:
      return tr->shuttle_incr[index];
    case KJS_JOG:
      return tr->jog[index];
    case KJS_KEY_UP:
      return tr->key_up[index];
    case KJS_KEY_DOWN:
    default:
      return tr->key_down[index];
    }
  }
  return NULL;
}

void
send_stroke_sequence(translation *tr, int kjs, int index)
{
  stroke *s;
  char key_name[100];
  int nkeys = 0;

  s = fetch_stroke(tr, kjs, index);
  if (s == NULL) {
    s = fetch_stroke(default_translation, kjs, index);
  }
  if (debug_keys && s) {
    switch (kjs) {
    case KJS_SHUTTLE:
      sprintf(key_name, "S%d", index-7);
      print_stroke_sequence(key_name, "", s);
      break;
    case KJS_SHUTTLE_INCR:
      sprintf(key_name, "I%s", (index>0)?"R":"L");
      print_stroke_sequence(key_name, "", s);
      break;
    case KJS_JOG:
      sprintf(key_name, "J%s", (index>0)?"R":"L");
      print_stroke_sequence(key_name, "", s);
      break;
    case KJS_KEY_UP:
      sprintf(key_name, "K%d", index);
      print_stroke_sequence(key_name, "U", s);
      break;
    case KJS_KEY_DOWN:
    default:
      sprintf(key_name, "K%d", index);
      print_stroke_sequence(key_name, "D", s);
      break;
    }
  }
  while (s) {
    if (s->keysym) {
      send_key(s->keysym, s->press);
      nkeys++;
#if HAVE_JACK
    } else {
      send_midi(s->status, s->data, s->incr, kjs, index);
#endif
    }
    s = s->next;
  }
  // no need to flush the display if we didn't send any keys
  if (nkeys) {
    XFlush(display);
  }
}

void
key(unsigned short code, unsigned int value, translation *tr)
{
  code -= EVENT_CODE_KEY1;

  if (code <= NUM_KEYS) {
    send_stroke_sequence(tr, value ? KJS_KEY_DOWN : KJS_KEY_UP, code);
  } else {
    fprintf(stderr, "key(%d, %d) out of range\n", code + EVENT_CODE_KEY1, value);
  }
}


void
shuttle(int value, translation *tr)
{
  if (value < -7 || value > 7) {
    fprintf(stderr, "shuttle(%d) out of range\n", value);
  } else {
    gettimeofday(&last_shuttle, 0);
    need_synthetic_shuttle = value != 0;
    if( value != shuttlevalue ) {
      if (shuttlevalue < -7 || shuttlevalue > 7) {
	// not yet initialized, assume 0
	shuttlevalue = 0;
      }
      int direction = (value < shuttlevalue) ? -1 : 1;
      int index = direction > 0 ? 1 : 0, val = value+7;
      if (fetch_stroke(tr, KJS_SHUTTLE_INCR, index)) {
	while (shuttlevalue != value) {
	  send_stroke_sequence(tr, KJS_SHUTTLE_INCR, index);
	  shuttlevalue += direction;
	}
      } else
	shuttlevalue = value;
      send_stroke_sequence(tr, KJS_SHUTTLE, val);
    }
  }
}

// Due to a bug (?) in the way Linux HID handles the ShuttlePro, the
// center position is not reported for the shuttle wheel.  Instead,
// a jog event is generated immediately when it returns.  We check to
// see if the time since the last shuttle was more than a few ms ago
// and generate a shuttle of 0 if so.
//
// Note, this fails if jogvalue happens to be 0, as we don't see that
// event either!
void
jog(unsigned int value, translation *tr)
{
  int direction;
  struct timeval now;
  struct timeval delta;

  // We should generate a synthetic event for the shuttle going
  // to the home position if we have not seen one recently
  if (need_synthetic_shuttle) {
    gettimeofday( &now, 0 );
    timersub( &now, &last_shuttle, &delta );

    if (delta.tv_sec >= 1 || delta.tv_usec >= 5000) {
      shuttle(0, tr);
      need_synthetic_shuttle = 0;
    }
  }

  if (jogvalue != 0xffff) {
    value = value & 0xff;
    direction = ((value - jogvalue) & 0x80) ? -1 : 1;
    while (jogvalue != value) {
      // driver fails to send an event when jogvalue == 0
      if (jogvalue != 0) {
	send_stroke_sequence(tr, KJS_JOG, direction > 0 ? 1 : 0);
      }
      jogvalue = (jogvalue + direction) & 0xff;
    }
  }
  jogvalue = value;
}

void
jogshuttle(unsigned short code, unsigned int value, translation *tr)
{
  switch (code) {
  case EVENT_CODE_JOG:
    jog(value, tr);
    break;
  case EVENT_CODE_SHUTTLE:
    shuttle(value, tr);
    break;
  default:
    fprintf(stderr, "jogshuttle(%d, %d) invalid code\n", code, value);
    break;
  }
}

char *
get_window_name(Window win)
{
  Atom prop = XInternAtom(display, "WM_NAME", False);
  Atom type;
  int form;
  unsigned long remain, len;
  unsigned char *list;

  if (XGetWindowProperty(display, win, prop, 0, 1024, False,
			 AnyPropertyType, &type, &form, &len, &remain,
			 &list) != Success) {
    fprintf(stderr, "XGetWindowProperty failed for window 0x%x\n", (int)win);
    return NULL;
  }

  return (char*)list;
}

char *
get_window_class(Window win)
{
  Atom prop = XInternAtom(display, "WM_CLASS", False);
  Atom type;
  int form;
  unsigned long remain, len;
  unsigned char *list;

  if (XGetWindowProperty(display, win, prop, 0, 1024, False,
			 AnyPropertyType, &type, &form, &len, &remain,
			 &list) != Success) {
    fprintf(stderr, "XGetWindowProperty failed for window 0x%x\n", (int)win);
    return NULL;
  }

  return (char*)list;
}

char *
walk_window_tree(Window win, char **window_class)
{
  char *window_name;
  Window root = 0;
  Window parent;
  Window *children;
  unsigned int nchildren;

  while (win != root) {
    window_name = get_window_name(win);
    if (window_name != NULL) {
      *window_class = get_window_class(win);
      return window_name;
    }
    if (XQueryTree(display, win, &root, &parent, &children, &nchildren)) {
      win = parent;
      XFree(children);
    } else {
      fprintf(stderr, "XQueryTree failed for window 0x%x\n", (int)win);
      return NULL;
    }
  }
  return NULL;
}

static Window last_focused_window = 0;
static translation *last_window_translation = NULL;

translation *
get_focused_window_translation()
{
  Window focus;
  int revert_to;
  char *window_name = NULL, *window_class = NULL;
  char *name;

  XGetInputFocus(display, &focus, &revert_to);
  if (focus != last_focused_window) {
    last_focused_window = focus;
    window_name = walk_window_tree(focus, &window_class);
    if (window_name == NULL) {
      name = "-- Unlabeled Window --";
    } else {
      name = window_name;
    }
    last_window_translation = get_translation(name, window_class);
    if (debug_regex) {
      if (last_window_translation != NULL) {
	printf("translation: %s for %s (class %s)\n",
	       last_window_translation->name, name, window_class);
      } else {
	printf("no translation found for %s (class %s)\n", name, window_class);
      }
    }
    if (window_name != NULL) {
      XFree(window_name);
    }
    if (window_class != NULL) {
      XFree(window_class);
    }
  }
  return last_window_translation;
}

void
handle_event(EV ev)
{
  translation *tr = get_focused_window_translation();
  
  //fprintf(stderr, "event: (%d, %d, 0x%x)\n", ev.type, ev.code, ev.value);
  if (tr != NULL) {
    switch (ev.type) {
    case EVENT_TYPE_DONE:
    case EVENT_TYPE_ACTIVE_KEY:
      break;
    case EVENT_TYPE_KEY:
      key(ev.code, ev.value, tr);
      break;
    case EVENT_TYPE_JOGSHUTTLE:
      jogshuttle(ev.code, ev.value, tr);
      break;
    default:
      fprintf(stderr, "handle_event() invalid type code\n");
      break;
    }
  }
}

void help(char *progname)
{
  fprintf(stderr, "Usage: %s [-h] [-o] [-p] [-j name] [-r rcfile] [-d[rskj]] [device]\n", progname);
  fprintf(stderr, "-h print this message\n");
  fprintf(stderr, "-o enable MIDI output\n");
  fprintf(stderr, "-p enable hot-plugging\n");
  fprintf(stderr, "-j jack client name (default: shuttlepro)\n");
  fprintf(stderr, "-r config file name (default: SHUTTLE_CONFIG_FILE variable or ~/.shuttlerc)\n");
  fprintf(stderr, "-d debug (r = regex, s = strokes, k = keys, j = jack; default: all)\n");
  fprintf(stderr, "device, if specified, is the name of the shuttle device to open.\n");
  fprintf(stderr, "Otherwise the program will try to find a suitable device on its own.\n");
}

void jack_warning(char *progname)
{
  fprintf(stderr, "%s: Warning: this version was compiled without Jack support\n", progname);
  fprintf(stderr, "Try recompiling the program with Jack installed to enable this option.\n");
}

#include <glob.h>

// Helper functions to process the command line, so that we can pass it to
// Jack session management.

static char *command_line;
static size_t len;

static void add_command(char *arg)
{
  char *a = arg;
  // Do some simplistic quoting if the argument contains blanks. This won't do
  // the right thing if the argument also contains quotes. Oh well.
  if ((strchr(a, ' ') || strchr(a, '\t')) && !strchr(a, '"')) {
    a = malloc(strlen(arg)+3);
    sprintf(a, "\"%s\"", arg);
  }
  if (!command_line) {
    len = strlen(a);
    command_line = malloc(len+1);
    strcpy(command_line, a);
  } else {
    size_t l = strlen(a)+1;
    command_line = realloc(command_line, len+l+1);
    command_line[len] = ' ';
    strcpy(command_line+len+1, a);
    len += l;
  }
  if (a != arg) free(a);
}

static char *absolute_path(char *name)
{
  if (*name == '/') {
    return name;
  } else {
    // This is a relative pathname, we turn it into a canonicalized absolute
    // path.  NOTE: This requires glibc. We should probably rewrite this code
    // to be more portable.
    char *pwd = getcwd(NULL, 0);
    if (!pwd) {
      perror("getcwd");
      return name;
    } else {
      char *path = malloc(strlen(pwd)+strlen(name)+2);
      static char abspath[PATH_MAX];
      sprintf(path, "%s/%s", pwd, name);
      if (!realpath(path, abspath)) strcpy(abspath, path);
      free(path); free(pwd);
      return abspath;
    }
  }
}

uint8_t quit = 0;

void quit_callback()
{
  quit = 1;
}

int
main(int argc, char **argv)
{
  EV ev;
  int nread;
  char *dev_name;
#if HAVE_JACK
  char *client_name = "shuttlepro";
#endif
  int fd;
  int first_time = 1;
  int opt, hotplug = 0;

  // Start recording the command line to be passed to Jack session management.
  add_command(argv[0]);

  while ((opt = getopt(argc, argv, "hopd::j:r:")) != -1) {
    switch (opt) {
    case 'h':
      help(argv[0]);
      exit(0);
    case 'p':
      hotplug = 1;
      add_command("-p");
      break;
    case 'o':
#if HAVE_JACK
      enable_jack = 1;
      add_command("-o");
#else
      jack_warning(argv[0]);
#endif
      break;
    case 'd':
      if (optarg && *optarg) {
	const char *a = optarg; char buf[100];
	snprintf(buf, 100, "-d%s", optarg);
	add_command(buf);
	while (*a) {
	  switch (*a) {
	  case 'r':
	    default_debug_regex = 1;
	    break;
	  case 's':
	    default_debug_strokes = 1;
	    break;
	  case 'k':
	    default_debug_keys = 1;
	    break;
	  case 'j':
	    debug_jack = 1;
	    break;
	  default:
	    fprintf(stderr, "%s: unknown debugging option (-d), must be r, s, k or j\n", argv[0]);
	    fprintf(stderr, "Try -h for help.\n");
	    exit(1);
	  }
	  ++a;
	}
      } else {
	default_debug_regex = default_debug_strokes = default_debug_keys = 1;
	debug_jack = 1;
	add_command("-d");
      }
      break;
    case 'j':
#if HAVE_JACK
      client_name = optarg;
      add_command("-j");
      add_command(optarg);
#else
      jack_warning(argv[0]);
#endif
      break;
    case 'r':
      config_file_name = optarg;
      add_command("-r");
      // We need to convert this to an absolute pathname for Jack session
      // management.
      add_command(absolute_path(optarg));
      break;
    default:
      fprintf(stderr, "Try -h for help.\n");
      exit(1);
    }
  }

  if (optind+1 < argc) {
    help(argv[0]);
    exit(1);
  }

  if (optind >= argc) {
    // Try to find a suitable device. The following glob pattern should
    // hopefully cover all existing versions.
    glob_t globbuf;
    if (glob("/dev/input/by-id/usb-*Shuttle*-event-if*",
	     0, NULL, &globbuf)) {
      fprintf(stderr, "%s: found no suitable shuttle device\n", argv[0]);
      fprintf(stderr, "Please make sure that your shuttle device is connected.\n");
      fprintf(stderr, "You can also specify the device name on the command line.\n");
      fprintf(stderr, "Try -h for help.\n");
      exit(1);
    } else {
      dev_name = globbuf.gl_pathv[0];
      fprintf(stderr, "%s: found shuttle device:\n%s\n", argv[0], dev_name);
    }
  } else {
    dev_name = argv[optind];
  }

  initdisplay();

#if HAVE_JACK
  if (command_line) jack_command_line = command_line;
  if (enable_jack) {
    seq.client_name = client_name;
    if (!init_jack(&seq, debug_jack)) enable_jack = 0;
  }
#endif

  time_t last_t = time(NULL);
  struct sigaction int_handler = { .sa_handler = quit_callback };
  sigaction(SIGINT, &int_handler, 0);
  while (!quit) {
    fd = open(dev_name, O_RDONLY);
    if (fd < 0) {
      perror(dev_name);
      if (first_time) {
	exit(1);
      }
    } else {
      // Flag it as exclusive access
      if(ioctl( fd, EVIOCGRAB, 1 ) < 0) {
	perror( "evgrab ioctl" );
      } else {
	first_time = 0;
	while (!quit) {
	  nread = read(fd, &ev, sizeof(ev));
	  if (nread < 0 && (hotplug || quit)) {
	    quit = 1;
	    break;
	  }
	  // Check whether to reload the config file in regular intervals.
	  // Note that if the file *is* reloaded, then we also need to reset
	  // last_focused_window here, so that the translations of the focused
	  // window are recomputed the next time we handle an event.
	  time_t t = time(NULL);
	  if (t > last_t && read_config_file()) {
	    last_focused_window = 0;
	    last_t = t;
	  }
	  if (nread == sizeof(ev)) {
	    handle_event(ev);
	    // Make sure that debugging output gets flushed every once in a
	    // while (may be buffered when we're running inside a QjackCtl
	    // session).
	    if (debug_regex || debug_strokes || debug_keys)
	      fflush(stdout);
	  } else {
	    if (nread < 0) {
	      perror("read event");
	      break;
	    } else {
	      fprintf(stderr, "short read: %d\n", nread);
	      break;
	    }
	  }
	}
      }
    }
    close(fd);
    if (!quit) sleep(1);
  }
}
